sql/sql_connect.cc

```diff
@@ -1000,6 +1000,28 @@ static int check_connection(THD *thd)
     bzero((char*) &net->vio->remote, sizeof(net->vio->remote));
   }
   vio_keepalive(net->vio, TRUE);
+
+  if (global_system_variables.keepalive_probes != 9)
```

Comment:
That's better. However from looking at it, I think I a single vio function for all OSes is a better abstraction, as compared to a win32 function and another function for anything else. Something along the lines

struct vio_keepalive_opts{
   int interval;
   int idle;
   int probes;
};
vio_set_keepalive_options(Vio * vio, const vio_keepalive_opts *opts)

On some OS it would amount to  at most 3 setsockopt calls, On Windows it would amount to single  WSAIoctl call .